File: Editor/src/Editor/Camera.cpp
#include <Yngin/Core/Context.h>
#include <Yngin/Core/InputSystem.h>
#include "Editor.h"

using namespace Yngin;

void Editor::handleCameraMovement(Yngin::Camera* camera) {
	Context* ctx = camera->getContext();
	InputSystem* input = ctx->getInputSystem();

	double delta = ctx->getDeltaTime();

	ctx->getWindow()->setCursorLocked(input->isMousePressed(MOUSE_BUTTON::RIGHT));

	if (input->isMouseJustPressed(MOUSE_BUTTON::RIGHT)) {
		input->setMousePosition({ 0, 0 });
	}

	if (input->isMousePressed(MOUSE_BUTTON::RIGHT)) {
		glm::vec3 o = camera->getOrientation();
		glm::ivec2 m = input->getMousePosition(true);
		input->setMousePosition({ 0, 0 });

		float senstivity = 0.002f;

		float yaw = atan2(o.x, o.y);
		float pitch = asin(o.z);

		o.x = cos(pitch) * sin(yaw + m.x * senstivity);
		o.y = cos(pitch) * cos(yaw + m.x * senstivity);
		o.z = sin(pitch - m.y * senstivity);

		camera->setOrientation(o);

		glm::vec3 forward = glm::normalize(o);
		glm::vec3 right = glm::normalize(glm::cross(forward, glm::vec3(0, 0, 1)));
		glm::vec3 realUp = glm::cross(forward, right);

		glm::vec3 change = {};

		if (input->isKeyPressed(Yngin::KEY::W)) {
			change += forward;
		}
		if (input->isKeyPressed(Yngin::KEY::S)) {
			change -= forward;
		}
		if (input->isKeyPressed(Yngin::KEY::D)) {
			change += right;
		}
		if (input->isKeyPressed(Yngin::KEY::A)) {
			change -= right;
		}
		if (input->isKeyPressed(Yngin::KEY::Q)) {
			change += realUp;
		}
		if (input->isKeyPressed(Yngin::KEY::E)) {
			change -= realUp;
		}

		float speed = 5.0f;
		if (input->isKeyPressed(Yngin::KEY::LSHIFT)) {
			speed *= 1.5f;
		}

		camera->setPosition(camera->getPosition() + change * (float)delta * speed);
	}
}

File: DemoApp/assets/camera.lua
local input = Yngin.InputSystem
local window = Yngin.Window

function onUpdate(dt)
	window:setCursorLocked(input:isMousePressed(MOUSE_BUTTON.RIGHT))
	
	if (input:isMouseJustPressed(MOUSE_BUTTON.RIGHT)) then
		input:setMousePosition(IVec2.new(0))
	end
	
	local camera = Yngin.ScenesManager:getActive():getCamerasManager():getCamera(0)

	if (input:isMousePressed(MOUSE_BUTTON.RIGHT)) then
		local o = camera:getOrientation()
		local m = input:getMousePosition(true)

		local senstivity = 0.002

		local yaw = math.atan2(o.x, o.y)
		local pitch = math.asin(o.z)

		o.x = math.cos(pitch) * math.sin(yaw + m.x * senstivity)
		o.y = math.cos(pitch) * math.cos(yaw + m.x * senstivity)
		o.z = math.sin(math.max(math.min(pitch - m.y * senstivity, 1.57), -1.57))

		camera:setOrientation(o)

		input:setMousePosition(IVec2.new())

		local forward = o:normalize()
		local right = forward:cross(Vec3.new(0, 0, 1)):normalize()
		local realUp = forward:cross(right)

		local change = Vec3.new()

		if (input:isKeyPressed(KEY.W)) then
			change = change + forward
		end
		if (input:isKeyPressed(KEY.S)) then
			change = change - forward
		end
		if (input:isKeyPressed(KEY.D)) then
			change = change + right
		end
		if (input:isKeyPressed(KEY.A)) then
			change = change - right
		end
		if (input:isKeyPressed(KEY.Q)) then
			change = change + realUp
		end
		if (input:isKeyPressed(KEY.E)) then
			change = change - realUp
		end

		local speed = 5.0
		if (input:isKeyPressed(KEY.LSHIFT)) then
			speed = speed * 1.5
		end

		camera:setPosition(camera:getPosition() + change * dt * speed)
	end
end
